Source/WebCore/platform/graphics/ImageFrameWorkQueue.h

```diff
@@ -0,0 +1,90 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "DecodingOptions.h"
+#include "ImageTypes.h"
+#include <wtf/SynchronizedFixedQueue.h>
+#include <wtf/ThreadSafeRefCounted.h>
+#include <wtf/WorkQueue.h>
+
+namespace WebCore {
+
+class ImageFrameProvider;
+
+class ImageFrameWorkQueue : public ThreadSafeRefCounted<ImageFrameWorkQueue> {
+public:
+    struct Request {
+        unsigned index;
+        SubsamplingLevel subsamplingLevel;
+        ImageAnimatingState animatingState;
+        DecodingOptions options;
+        friend bool operator==(const Request&, const Request&) = default;
+    };
+
+    static Ref<ImageFrameWorkQueue> create(ImageFrameProvider&);
+
+    void start();
+    void dispatch(const Request&);
+    void stop();
+
+    bool isIdle() const { return m_decodeQueue.isEmpty(); }
+
+    void didCacheNativeImageAtIndex(unsigned index);
+    void didDestroyNativeImageAtIndex(unsigned index);
+
+    bool isPendingDecodingAtIndex(unsigned index, SubsamplingLevel, const DecodingOptions&) const;
+    bool isPendingDrawingAtIndex(unsigned index) const;
+
+    void setMinimumDecodingDurationForTesting(Seconds duration) { m_minimumDecodingDurationForTesting = duration; }
+
+    void dump(TextStream&) const;
+
+private:
+    ImageFrameWorkQueue(ImageFrameProvider&);
+
+    static const int BufferSize = 8;
+    using RequestQueue = SynchronizedFixedQueue<Request, BufferSize>;
+    using DecodeQueue = Deque<Request, BufferSize>;
+    using CommitMap = HashMap<unsigned, MonotonicTime>;
+
+    RequestQueue& requestQueue();
+    DecodeQueue& decodeQueue() { return m_decodeQueue; }
+    CommitMap& commitMap() { return m_commitMap; }
+
+    Seconds minimumDecodingDurationForTesting() const { return m_minimumDecodingDurationForTesting; }
+
+    ImageFrameProvider& m_frameProvider;
```

Comment:
Same here, can we make this `ThreadSafeWeakPtr`? (Would be nice if we had a `ThreadSafeWeakRef`, just like we have `WeakRef` for weak pointers that we never expect to null when we access them.)

Reply:
This is fixed in #23318. But instead of `ImageFrameProvider`, `ImageSource` is used for now. This will change once `ImageFrameProvider` is introduced.